src/plugins/share/public/url_service/redirect/redirect_manager.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { CoreSetup } from 'src/core/public';
+import { i18n } from '@kbn/i18n';
+import { BehaviorSubject } from 'rxjs';
+import { migrateToLatest } from '../../../../kibana_utils/common';
+import type { SerializableState } from '../../../../kibana_utils/common';
+import type { UrlService } from '../../../common/url_service';
+import { render } from './render';
+import { parseSearchParams } from './util/parse_search_params';
+
+export interface RedirectOptions {
+  /** Locator ID. */
+  id: string;
+
+  /** Kibana version when locator params where generated. */
+  version: string;
+
+  /** Locator params. */
+  params: unknown & SerializableState;
+}
+
+export interface RedirectManagerDependencies {
+  url: UrlService;
+}
+
+export class RedirectManager {
+  public readonly error$ = new BehaviorSubject<null | Error>(null);
+
+  constructor(public readonly deps: RedirectManagerDependencies) {}
+
+  public registerRedirectApp(core: CoreSetup) {
+    core.application.register({
+      id: 'r',
+      title: 'Redirect endpoint',
+      chromeless: true,
+      mount: (params) => {
+        const unmount = render(params.element, { manager: this });
+        this.onMount(params.history.location.search);
+        return () => {
+          unmount();
+        };
+      },
+    });
+  }
+
+  public onMount(urlLocationSearch: string) {
+    const options = this.parseSearchParams(urlLocationSearch);
+    const locator = this.deps.url.locators.get(options.id);
+
+    if (!locator) {
+      const message = i18n.translate('share.urlService.redirect.RedirectManager.locatorNotFound', {
+        defaultMessage: 'Locator [ID = {id}] does not exist.',
+        values: {
+          id: options.id,
+        },
+        description:
+          'Error displayed to user in redirect endpoint when redirection cannot be performed successfully, because locator does not exist.',
+      });
+      const error = new Error(message);
+      this.error$.next(error);
+      throw error;
+    }
+
+    const { state: migratedParams } = migrateToLatest(locator.migrations, {
```

Comment:
cc @stacey-gammon  should we make migration happen internally inside navigate() ? one less thing that to worry about for the consumer, but not really. `navigate` would change and optionally accept version in. consumer would still need to make sure correct version is provided if the state is not of latest version. having to do them explicitly is kind of nice as you can check if anything changed in the state and warn the user about it.